numcodecs/compat.py

```diff
@@ -27,80 +26,161 @@
     from functools import reduce
 
 
-def buffer_tobytes(v):
-    """Obtain a sequence of bytes for the memory buffer used by `v`."""
-    if isinstance(v, binary_type):
-        return v
-    elif isinstance(v, np.ndarray):
-        return v.tobytes(order='A')
-    elif PY2 and isinstance(v, array.array):  # pragma: py3 no cover
-        return v.tostring()
-    else:
-        return memoryview(v).tobytes()
+def ensure_text(l, encoding='utf-8'):
+    if isinstance(l, text_type):
+        return l
+    else:  # pragma: py3 no cover
+        return text_type(l, encoding=encoding)
 
 
-def buffer_copy(buf, out=None):
-    """Copy the contents of the memory buffer from `buf` to `out`."""
+def ensure_ndarray(buf, dtype=None):
+    """Convenience function to coerce `buf` to a numpy array, if it is not already a
+    numpy array.
 
-    if out is None:
-        # no-op
-        return buf
+    Parameters
+    ----------
+    buf : array-like or bytes-like
+        A numpy array or any object exporting a buffer interface.
+    dtype : dtype, optional
+        Request that the data be viewed as the given dtype.
 
-    # handle ndarray destination
-    if isinstance(out, np.ndarray):
+    Returns
+    -------
+    arr : ndarray
+        A numpy array, sharing memory with `buf`.
 
-        # view source as destination dtype
-        if isinstance(buf, np.ndarray):
-            buf = buf.view(dtype=out.dtype).reshape(-1, order='A')
-        else:
-            buf = np.frombuffer(buf, dtype=out.dtype)
+    Notes
+    -----
+    This function will not create a copy under any circumstances, it is guaranteed to
+    return a view on memory exported by `buf`.
 
-        # ensure shapes are compatible
-        if buf.shape != out.shape:
-            if out.flags.f_contiguous:
-                order = 'F'
-            else:
-                order = 'C'
-            buf = buf.reshape(out.shape, order=order)
+    """
 
-        # copy via numpy
-        np.copyto(out, buf)
+    if isinstance(buf, np.ndarray):
+        # already a numpy array
+        arr = buf
+
+    elif isinstance(buf, array.array) and buf.typecode == 'u':
```

Comment:
Could we also do this for [`array`'s of type `"c"`]( https://docs.python.org/2/library/array.html#module-array )?

Reply:
I think arrays of type "c" actually work ok, come out as numpy type "S1". There's a test for this in test_compat.

Reply:
Good point, I've disallowed "c" in latest commit.